### ISSUE TITLE
[Feature Request]: Enhance .obsidian handling when adding subsequent clients to sync

### ISSUE THREAD

**Issue by schemen** [sboesen/remotely-sync#109]
### What feature are you suggesting?

When you add a subsequent client to sync, an issue arises when you sync settings as well. As it turns out, your new client will have "newer" settings files etc. This means, it'll erase all of the .obsidian settings coming from the other client which breaks many things of course. 

Ideally, you would get a "first time pull" sequence where you simply pull the entire remote data, maybe with the option to skip .obsidian or a force pull as well.

At the moment, I can't set up subsequent clients because it erases all of the customization I've made.

### What remote cloud services are you using, or suggesting adding the feature to?

webdav

### Ensure no sensitive information

- [X] I ensure that no sensitive information is submitted in the issue.

**Comment by kadisonm** [comment]
Haven't ran into this issue myself but I have ran into issues with deleting plugins.

Did you install all the plugins on your second client before running a sync? That might be why.

Try syncing on your second client (sync config dir enabled) without any plugins installed except Remotely Sync. This should install all the other plugins and settings from the first client.

Let me know if it is still happening or if the above doesn't apply.

**Comment by schemen** [comment]
Hi! Thanks for the reply

It was a clean install (New vault) with onyl Remotely Sync installed. On the main clients backup (Git) you could see that once the second client was done with syncing, it sadly pushed it's (newer and empty) config files over since they have a newer mtime than the primary clients config files. 

The plugins got pulled over (since they don't exist yet, thus they get copied), but the obsidian config that enables them is empty with the exception of Remotely Sync.

The config I'm talking about are things that obsidian creates upon vault creating, like app.json, appearance.json, community-plugins.json etc.

**Comment by kadisonm** [comment]
Oh apologies I misread. Just to confirm are your plugin settings sync correctly?

I have also found I have to redo a lot of vault settings on the initial sync. Thank you for getting to the root cause of it. 

I think the sync algorithm needs some discussion @sboesen since there is also issues with deleting.

Not sure if there is a quick fix to this other than syncing creation time and comparing them or something similar

**Comment by schemen** [comment]
Yes correct, plugin and obsidian setting sync are enabled.

I think a quickfix would be what's proposed up there, essentially a first time client (and maybe a button in the plugin settings) pulls all of the data (maybe leave out the remotely sync settings?)

**Comment by kadisonm** [comment]
I could see that working. Could just pull the .obsidian folder from the remote and override the local .obsidian

Might help fix deleted plugins getting reinstalled but that might be something seperate.

Hopefully Sboesen can look into it as well.

**Comment by sboesen** [comment]
Ooo interesting - that makes sense! Thanks for filing the issue and sorry about that. 

These files are inherently "newer" on the remote. 

I think we need:
- a list of affected files
  - app.json, etc
- a check for last sync being null. if so, skip uploading these files

This can probably be added to the isSkipItem function. I don't think this will be possible to fix JUST by syncing mtime, but that will of course prevent some unnecessary file upload/downloads.

**Comment by schemen** [comment]
Sounds good - other than the last sync being null an optional "force download and overwrite all" button wouldn't be bad from a debugging perspective

**Comment by kadisonm** [comment]
Wouldn't this not work in the case that:
- Vault 1 is created and synced. Last sync is null, so it skips uploading.
- Vault 2 is created and synced after. Skips uploading too.

Won't neither vaults have synced their .obsidian files? At least not until they both run a sync again. In that case it would fix it since the vault mTimes would be older than the remote.

**Comment by schemen** [comment]
> Vault 1 is created and synced. Last sync is null, so it skips uploading.
Probably a check whether the file is actually there and the sync is null, skip. If the file isn't there at all, it can be uploaded.

**Comment by kadisonm** [comment]
> Probably a check whether the file is actually there and the sync is null, skip. If the file isn't there at all, it can be uploaded.

Yeah that was my thoughts as well but I don't think that works if the 2nd vault is created after the 1st vault's sync. Since the mTime's of the 2nd vault will be later than the remote. Which will then overwrite the remote when the 2nd vault syncs for the second time.

Maybe if last sync is null and it's files are on the remote, then download from the remote. 

> "force download and overwrite all"

I think this would also be good for debugging.

**Comment by sboesen** [comment]
> "force download and overwrite all"
I think this would also be good for debugging.

Agreed! Let's open a new issue to track this.

BTW I didn't get a chance to do much testing on this one, but I think this commit should fix it.

**Comment by kadisonm** [comment]
I had a look and it looks like it should as well. I'lll give it a test later. Thanks.

**Comment by kadisonm** [comment]
Wow great job it worked well for me. First time I haven't had to re set up my note and folder creation locations.

Although, I did notice the graph didn't have my filters. Is there a way to sync this?

**Comment by sboesen** [comment]
Glad that worked for you!

> Although, I did notice the graph didn't have my filters. Is there a way to sync this?

Hmm - graph.json should download first on a new device. You created groups on one computer, synced that, and then couldn't see those filters on a new device?

**Comment by kadisonm** [comment]
> You created groups on one computer, synced that, and then couldn't see those filters on a new device?

No actually, I tried again on another device and the fix didn't work. I'm not sure why it worked the first time for me though. Or at least the first time only app.json synced.

**Comment by sboesen** [comment]
I wasn't able to reproduce this locally (didn't test too deeply), but reopening this issue as I see there is a gap with how I implemented it currently:

This only works if you set `sync .obsidian` on the new client and *then* sync. Otherwise you'll do a sync, it'll exclude all .obsidian files, then enable the setting, and the problem persists.

**Comment by sboesen** [comment]
Dropping the priority as there is now a workaround and this should cover the normal use case (definitely happy to test more if that isn't working)

**Comment by kadisonm** [comment]
For me the issue is still happening. I'm wondering if maybe my other devices aren't properly syncing the .obsidian. It's odd though since my plugins are still syncing.

**Comment by kadisonm** [comment]
Ohh checking the plugin settings, when I create a new vault the lastSynced is -1.

https://github.com/sboesen/remotely-sync/blob/3066536526af8296364e54a349acb68ff42e72c8/src/sync.ts#L1410

The check only checks for if it's undefined. Not sure if there is a type conversion thing going on where undefined is equal to -1 since I do notice it is == and not ===. But maybe this could be why it's not working for me.

**Comment by sboesen** [comment]
Bah, that makes sense! I don't know how it was "working" for me but I probably wasn't reproducing the issue properly.

[Here](https://github.com/sboesen/remotely-sync/blob/master/src/main.ts#L91) is where default settings are set, and yes lastSynced is set to -1. Good catch!

**Comment by sboesen** [comment]
Should be fixed in 0.4.43 - please let me know if this is still happening!

**Comment by kadisonm** [comment]
> Bah, that makes sense! I don't know how it was "working" for me but I probably wasn't reproducing the issue properly.

Aha no worries. I also thought it worked for me the first time which made me confused.

**Comment by kadisonm** [comment]
Doesn't seem to work for me. I think you forgot to commit so maybe the release wasn't the updated version?
If it still doesn't work after then I'll take a look but from what I remember it looked like it should.

**Comment by sboesen** [comment]
Hey sorry! Will check by tomorrow, sorry about the delay!

**Comment by kadisonm** [comment]
No worries, I've been pretty busy as well.

**Comment by sboesen** [comment]
Added debug logging in https://github.com/sboesen/remotely-sync/releases/tag/0.4.44

Check https://github.com/sboesen/remotely-sync/commit/8d890b8dcbf08e2065c0bc3ba590dc502a972c62

**Comment by kadisonm** [comment]
Hi thanks for adding the debugging. I'll see if I can get check what's happening tomorrow. Did try on my phone quickly and noticed the log gets spammed. I think it might be best to put it before the iterations.

**Comment by kadisonm** [comment]
After checking the logs it looks like the keys have '.obsidian/' in front of them. Which is why they can't be found in the skipKeys and aren't passing the if check. (I believe, haven't tested)

https://github.com/sboesen/remotely-sync/blob/c9ee488a3fe7635357f9f8ce608a7290634cd0b5/src/sync.ts#L1412

**Comment by sboesen** [comment]
Auto closed. This attempted fix is released as part of 0.4.45 - hopefully this is the one!
